README.md
OSECZ Logo
==========
This project contains a script that generates the project logo images.


Contents
--------
* [Development Environment](#development-environment)
* [Generate Logo](#generator-logo)
* [Script Usage](#script-usage)
* [License](#license)
* [Copyright](#copyright)


Development Environment
-----------------------
Python3 and Inkscape are required to generate the logos. On Linux, enter
the following commands to install these tools.

    apt-get install make python3 inkscape

On Mac, enter the following commands.

    brew install python3
    brew cask install xquartz inkscape


Generate Logo
-------------
To generate the logo image files, enter the following command:

    make logo

The above command invokes the Python script `logo.py` and generates the
logo as SVG images. Then it invokes `inkscape` to convert the SVGs to
PNGs.

The output image files are written to the `out` directory.


Script Usage
------------
The script is a really simple Python script that accepts three arguments.

  - R: Radius of the red circle.
  - W: Width of each circle.
  - M: Additional space around the red circle.

The inner circle of the red band has a radius of R - W/2 and the outer
circle of the red band has a radius of R + W/2. In other words, the red
circle that is drawn is equivalent to drawing a red circle in the
following manner: First draw a red circle of radius R. Then make this
circle equally thick on both sides (inner as well as outer side) such
that the difference between the outer radius and the inner radius of
this circle is W. The final image is of size R + W/2 + M with a
transparent background.


License
-------
This is free and open source software. You can use, copy, modify,
merge, publish, distribute, sublicense, and/or sell copies of it,
under the terms of the MIT License. See [LICENSE.md][L] for details.

This software is provided "AS IS", WITHOUT WARRANTY OF ANY KIND,
express or implied. See [LICENSE.md][L] for details.

This license is applicable only to the source code in this project.
This license is not applicable to the image files generated by this
project.

[L]: LICENSE.md


Copyright
---------
Copyright (c) 2017-2018 OSECZ.org  
All Rights Reserved.

This copyright notice applies to all images generated by this project.


CHANGES.md
Changelog
=========

1.0.0 (2018-06-25)
------------------
### Added
- Logo with main colors: #ec0304, #028902, #166bff. These colors have
  luminance of 0.1791 approximately and contrast ratio of 4.58
  approximately with both #000000 and #ffffff.
- Logo with bright colors: #ff6657, #02b102, #129dff. These colors have
  luminance of 0.3146 approximately and contrast ratio of 4.58
  approximately with #303030.


logo.py
#!/usr/bin/env python3

import sys

# Read command line arguments
if len(sys.argv) == 7:
    r, t, m, c_r, c_g, c_b = sys.argv[1:]
else:
    print('Usage:', sys.argv[0],
         'RADIUS THICKNESS MARGIN RED GREEN BLUE', file=sys.stderr)
    sys.exit(1)

r, t, m = int(r), int(t), int(m)

# Validate radius and width
if r % 3 != 0:
    print('Error: RADIUS must be a multiple of 3', file=sys.stderr)
    sys.exit(1)

if t % 2 != 0:
    print('Error: WIDTH must be a multiple of 2', file=sys.stderr)
    sys.exit(1)

r_max = int(r + t/2 + m)
size = int(2 * r_max)
r_b = int(r / 3)
r_g = int(2 * r_b)
r_r = int(r)

print('size: {size}x{size}'.format(size=size), file=sys.stderr)
print('r:', int(r_r - t/2), r_r, int(r_r + t/2), file=sys.stderr)
print('g:', int(r_g - t/2), r_g, int(r_g + t/2), file=sys.stderr)
print('b:', int(r_b - t/2), r_b, int(r_b + t/2), file=sys.stderr)

logo = """<?xml version="1.0" encoding="UTF-8" ?>
<svg xmlns="http://www.w3.org/2000/svg" viewBox="-{r_max} -{r_max} {size} {size}" width="{size}" height="{size}">
<circle r="{r_r}" stroke="{c_r}" stroke-width="{t}" fill-opacity="0.0"/>
<circle r="{r_g}" stroke="{c_g}" stroke-width="{t}" fill-opacity="0.0"/>
<circle r="{r_b}" stroke="{c_b}" stroke-width="{t}" fill-opacity="0.0"/>
</svg>
"""
logo = logo.format(r_max=r_max, size=size,
                   r_r=r_r, r_g=r_g, r_b=r_b, t=t,
                   c_r=c_r, c_g=c_g, c_b=c_b)
print(logo)


colors.py
#!/usr/bin/env python3

"""Analyze luminance and contrast ratios of colors."""

# Reference: https://www.w3.org/TR/WCAG20-TECHS/G17.html#G17-tests


from math import sqrt


def srgb_to_linear(c):
    """Convert SRGB value of a color channel to linear value."""
    assert 0 <= c <= 1
    if c <= 0.03928:
        return c /12.92
    else:
        return ((c + 0.055) / 1.055)**2.4


def luminance(color_code):
    """Compute luminance of specified color."""
    # Scale the hex value of each channel to a value between 0 and 1.
    r = int(color_code[1:3], 16) / 255
    g = int(color_code[3:5], 16) / 255
    b = int(color_code[5:7], 16) / 255

    # Convert SRGB values to linear values.
    r = srgb_to_linear(r)
    g = srgb_to_linear(g)
    b = srgb_to_linear(b)

    # Luminance.
    return (0.2126 * r + 0.7152 * g + 0.0722 * b)


def contrast(L1, L2):
    """Compute contrast between two colors and/or luminance values."""
    if isinstance(L1, str):
        L1 = luminance(L1)

    if isinstance(L2, str):
        L2 = luminance(L2)

    if L1 < L2:
        L1, L2 = L2, L1

    a = 0.05
    return (L1 + a) / (L2 + a)


def analyze_colors(fg_colors, bg_colors):
    """Print luminance and contrast details of specified colors."""
    # Print header.
    print('{:7}  {:>6}'.format('', 'Lum'), end='')
    for bg in bg_colors:
        print('  {:7}'.format(bg), end='')
    print()

    # Print computation.
    for fg in fg_colors:
        print('{:7}  {:6.4f}'.format(fg, luminance(fg)), end='')
        for bg in bg_colors:
            print('  {:7.2f}'.format(contrast(fg, bg)), end='')
        print()


def html_preview(fg_colors, bg_colors):
    """Generate an HTML to preview the specified colors."""
    html = [
        '<!DOCTYPE html>',
        '<html lang="en">',
        '<head>',
        '<meta charset="UTF-8">',
        '<title>HTML Preview of Colors</title>',
        '<style>',
        'body { background: #c0c0c0; font-family: monospace; line-height: 1.5em }',
        'body > div { margin: 1em; padding: 1em; max-width: 15em }',
        'div > div { margin: 1em; text-align: center }',
        '</style>',
        '</head>',
        '<body>'
    ]

    for bg in bg_colors:
        html.append('')
        html.append('<div style="background: {}">'.format(bg))
        for fg in fg_colors:
            html.append('  <div style="background: {}">{}</div>'.format(fg, fg))
        html.append('</div>')

    html.append('')
    html.append('</body>')
    html.append('</html>')

    preview_filename = 'color-preview.html'
    with open(preview_filename, 'w') as f:
        f.write('\n'.join(html))
    print()
    print('Preview written to:', preview_filename)


def neutral_luminance(c1, c2):
    """Compute luminance with equal contrast with c1 and c2 (color codes)."""
    a = 0.05
    Lc1 = luminance(c1)
    Lc2 = luminance(c2)
    Ld, Lb = (Lc1, Lc2) if Lc1 <= Lc2 else (Lc2, Lc1)
    Lc = sqrt(a*a + Lb*a + Ld*a + Lb*Ld) - a
    return Lc


def desired_luminance(desired_contrast_ratio, color):
    """Luminance required to obtain desired contrast with color."""
    a = 0.05
    Lo = luminance(color)
    Lc = desired_contrast_ratio * (Lo + a) - a
    return Lc


def main():
    """Print ideal luminance and analyze details of chosen colors."""
    # Luminance of background colors we will work with.
    L_black = luminance('#000000')
    L_gray3 = luminance('#303030')
    L_white = luminance('#ffffff')
    print('Luminance: black: {:6.4f}'.format(L_black))
    print('Luminance: gray3: {:6.4f}'.format(L_gray3))
    print('Luminance: white: {:6.4f}'.format(L_white))
    print()

    # Contrast between background colors.
    print('Contrast: white/black: {:7.4f}'
          .format(contrast('#ffffff', '#000000')))
    print('Contrast: white/gray8: {:7.4f}'
          .format(contrast('#ffffff', '#808080')))
    print('Contrast: white/gray3: {:7.4f}'
          .format(contrast('#ffffff', '#303030')))
    print()

    # Luminance for neutral colors that have equal contrast with two
    # background colors.
    L_black_white_neutral = neutral_luminance('#000000', '#ffffff')
    L_gray3_white_neutral = neutral_luminance('#303030', '#ffffff')
    print('Neutral luminance: black/white: {:6.4f}'
          .format(L_black_white_neutral))
    print('Contrast with black: {:7.4f}'
          .format(contrast(L_black_white_neutral, L_black)))
    print('Contrast with white: {:7.4f}'
          .format(contrast(L_black_white_neutral, L_white)))
    print()
    print('Neutral luminance: gray3/white: {:6.4f}'
          .format(L_gray3_white_neutral))
    print('Contrast with gray3: {:7.4f}'
          .format(contrast(L_gray3_white_neutral, L_gray3)))
    print('Contrast with white: {:7.4f}'
          .format(contrast(L_gray3_white_neutral, L_white)))
    print()

    # Luminance of a bright color that has good contrast with gray3
    # background.
    L_desired = desired_luminance(4.5826, '#303030')
    print('Desired luminance: gray3: {:6.4f}'.format(L_desired))
    print('Contrast with gray3: {:7.4f}'.format(contrast(L_desired, L_gray3)))
    print()

    # Analyze the contrast of chosen foreground colors against chosen
    # background colors.
    fg_colors=['#ec0304', '#028902', '#166bff',
               '#ff6657', '#02b102', '#129dff']
    bg_colors=['#000000', '#303030', '#ffffff']
    analyze_colors(fg_colors, bg_colors)
    html_preview(fg_colors, bg_colors)


if __name__ == '__main__':
    main()
